src/core/thread/meshcop_dataset_manager.cpp

```diff
@@ -299,14 +300,76 @@ void DatasetManager::HandleSet(Coap::Header &aHeader, Message &aMessage, const I
             }
         }
 
+        if (tlvType == Tlv::kActiveTimestamp || tlvType == Tlv::kPendingTimestamp)
+        {
+            OT_TOOL_PACKED_BEGIN
+            struct
+            {
+                Tlv tlv;
+                Timestamp timestamp;
+            } OT_TOOL_PACKED_END timestampTlv;
+
+            timestampTlv.tlv.SetType(tlvType);
+            timestampTlv.tlv.SetLength(sizeof(Timestamp));
+            aMessage.Read(offset + sizeof(Tlv), sizeof(timestampTlv.timestamp), &timestampTlv.timestamp);
+            local.Set(timestampTlv.tlv);
+        }
+        else if (tlvType == Tlv::kChannelMask)
+        {
+            OT_TOOL_PACKED_BEGIN
```

Comment:
I don't think you need to duplicate the `Read()` - `Set()` code for each TLV type.  You could have a single code block that reads a generic TLV and sets the value in local set.